src/Core/src/HttpClient/AwsHttpClientFactory.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace AsyncAws\Core\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\HttpClient\RetryableHttpClient;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class AwsHttpClientFactory
+{
+    public static function createRetryableClient(HttpClientInterface $httpClient = null, LoggerInterface $logger = null): HttpClientInterface
+    {
+        if (null === $httpClient) {
+            $httpClient = HttpClient::create();
+        }
+        if (class_exists(RetryableHttpClient::class)) {
+            /** @psalm-suppress MissingDependency */
+            $httpClient = new RetryableHttpClient(
+                $httpClient,
+                new AwsRetryStrategy(),
```

Comment:
But how do we know what error factory to use?

Reply:
The default factory uses the chainFactory that iterate over known factories.

It's not as performant as the strategy generated by the client. But given it's only used in case of failure, I think it's enough.

Reply:
Maybe, we could move this HttpClie tFactory in the client... 🤔

Reply:
Yeh, that would work.

Reply:
I was thinking about this, but I feel uncomfortable with adding a `public static function createHttpClient` method in the abstractApi. It feels very hacky and breaks the separation of concerns...

PLUS, building specialized HttpClient could be an issue for people creating the HttpClient once and injecting it in multiple Services (ie S3Client and DynamoDb does not expect the same AWSErrorFactory)

Finally, I think it's better to
- do not inject anything and let the AbstractApi building it's own client.
- if people want to customize the HttpClient AND want the Retry thing expose a factory in a dedicated service that return a generic RetryStrategy that work with all the services.

Reply:
I would be fine with just the original PR that was accidentality merged yesterday. I already have the code to create the strategy anyway, for each of the services I need.

Reply:
> do not inject anything and let the AbstractApi building it's own client.

I definitely do not want that. I need to share the client to avoid blowing through my open file limit of 1000 on aws lambda. curl doesn't reuse - i found that out the hard way...

Reply:
I think we agreed. (I also want letting people optionally injecting their own HttpClient)

See the current diff of this PR (which is closed to the first version of this PR + exposes an optional `AwsHttpClientFactory` if people need help to create a Client.

The AbstractApi takes an optional HttpClientInterface
- People that want the default behavior just use `new S3Client()`, and the AbstractApi creates a `RetryableHttpClient`
- People that want full control can inject their own HttpClient `new S3Clien(.., $myClient)`, and the AbstractApi does nothing
- People that want Retryable thing with their own httpclient can use the new Factory `new S3Clien(.., AwsHttpClientFactory::createRetryableClient($myClient))`

Reply:
yeh